Add release notes for 0.26 (#529)
apple/pkl: docs/modules/release-notes/pages/0.26.adoc

Pkl is now available on Windows!

In addition to macOS and Linux, Pkl is now available for download on Windows operating systems.

For installation instructions, consult xref:pkl-cli:index.adoc#windows-executable[Windows Executable] in the CLI reference.

==== Writing cross-platform Pkl programs

On Windows, the in-language filepath separator is still `/`.
This means that most existing Pkl code should still continue working as-is.

A small portion of programs will need to be updated to handle Windows support correctly.
One example is the various conversion scripts that read input files as CLI flags.

For example, here is a snippet from module https://github.com/apple/pkl-pantry/blob/d9571cdce407bb7f1687f81b945a4bfd24ce017d/packages/k8s.contrib/convert.pkl#L100C1-L103C74[k8s.contrib.convert]:

.convert.pkl#L100-L103
[source,pkl]
----
local inputUri =
  if (input.startsWith(Regex(#"\w+:"#))) input      // absolute URI
  else if (input.startsWith("/")) "file://\(input)" // absolute file path
  else "file://\(read("env:PWD"))/\(input)"         // relative file path
----

On Windows, the env var `PWD` is a Windows-style path, like `C:\Foo\Bar\Baz`.

> With the above logic, `inputUri` can expand to a value like `"\file://C:\Foo\Bar\Baz/input.yaml"`.

Author: See below--the first `\` is just there to disable autolinking.

> This would an invalid URI; the correct URI should be `"\file:///C:/Foo/Bar/Baz/input.yaml"`.

Author: It's actually not a form feed. Rather, it disables asciidoc's autolinking.

Typealiases are meant to be statically defined, and shouldn't be able to see properties that can be late-bound.
Thus, a new rule is introduced: a typealias can only reference properties/methods that are `const`.

This is a breaking change. To read more about this breakage and remediation steps, reference <<typealiases-are-treated-as-const>>.

== Noteworthy [small]#🎶#

Ready when you need them.

[[pkldoc-improvements]]
=== Pkldoc improvements

The following improvements have been made to xref:pkl-doc:index.adoc[Pkldoc]:

1. The documentation for the standard library no longer shows dependent packages, because every package implicitly depends on the standard library (https://github.com/apple/pkl/pull/503[#503]).
2. If the standard library exists in the package list, it is shown first (https://github.com/apple/pkl/pull/165[#165]).
3. Fewer characters are percent-encoded. For example, the `(` and `)` characters are no longer percent-encoded (https://github.com/apple/pkl/pull/489[#489]).

[[standard-library-changes]]
=== Standard library changes (https://github.com/apple/pkl/pull/506[#506])

Multiple changes have been more to the standard library.

==== New property in `pkl.settings`

> Module `pkl.settings` has new property `http: Http?`.

Author: This section is meant to list all the changes made to the stdlib, so this doens't really seem out of place to me. But to help with the disjointed-ness, I'll add a section to this to link to the HTTP proxy section in the noteworthy. Also, changed this to bullet points instead of sub-sections.

Things to watch out for when upgrading.

[[typealiases-are-treated-as-const]]
=== Typealiases are treated as `const` (https://github.com/apple/pkl/pull/516[#516])

Typealiases are types that can stand in for another type declaration.

The aliased type can have constraints, where these constraints can reference values defined on the enclosing module.

.baseModule.pkl
[source,pkl]
----
typealias MyValue = Any(isValid) // <1>

isValid = true
----
<1> `isValid` is defined on the enclosing module.

One problem with this is that typealiases are meant to be statically defined.
Like classes, typealiases should not be changed by amending its enclosing module.

[source,pkl]
----
amends "baseModule.pkl"

isValid = false // <1>
----
<1> Despite this amended value, `typealias MyValue` should still stand for `Any(true)`.

To have clearer semantics, a new rule is introduced: referenced members on the enclosing module must be `const`.
Effectively, a typealias is treated as if it is also a `const` member.

To fix the above typealias, the `const` modifier should be added to `isValid`.

[source,diff]
----
 typealias MyValue = Any(isValid)
 
-isValid = true
+const isValid = true
----

> If the property can't be marked as `const` (maybe because it is amended somewhere downstream), another fix is to self-import the enclosing module.

Author: Reworded this a bit

If Pkl's dependencies are vendored within a repository, these dependencies will need to be re-vendored.

=== Java API Breaking Changes

The following Java APIs have breaking changes:

|===
|Class/method | Breaking change

|`org.pkl.cli.commands.CliDownloadPackageCommand`
|Renamed to `org.pkl.cli.commands.CliPackageDownloader`

|`org.pkl.cli.commands.CliAbstractProjectCommand`
|Renamed to `org.pkl.cli.commands.CliProjectCommand`

|`org.pkl.config.java.InvalidMappingException`
|Fields `pklName` and `javaName` are now `private`.

|`org.pkl.core.packages.PackageResolver#getInstance`
|New parameter `httpClient` added

|`org.pkl.core.repl.ReplServer.ReplServer`
|New parameter `httpClient` added

|`org.pkl.commons.cli.CliBaseOptions`
|New parameters `testPort`, `httpProxy`, `httpNoProxy` added
|===

[[standard-library-breaking-changes]]
=== Standard library breaking changes (https://github.com/apple/pkl/pull/506[#506])

Class `pkl.Project.EvaluatorSettings` has been removed.
A new (deprecated) typealias is added that points to new module `pkl.EvaluatorSettings`.

[[type-checked-settings-file]]
=== Type-checked settings file (https://github.com/apple/pkl/pull/477[#477])

> The xref:pkl-cli:index.adoc#settings-file[Settings file] is a way to control the behavior of the Pkl CLI.
> This module is expected to `amends "pkl:settings"`, but this behavior is not checked.
> 
> In 0.26, it will be an error if the settings module does not amend `"pkl:settings"`, nor set its `output.value` to an instance of the settings module.

Reviewer: Other points are in past tense
```suggestion
The xref:pkl-cli:index.adoc#settings-file[Settings file] is a way to control the behavior of the Pkl CLI.
This module is expected to `amends "pkl:settings"`, but this behavior was not checked.

In 0.26, it is an error if the settings module neither amends `"pkl:settings"`, nor set its `output.value` to an instance of the settings module.
```

* `const` members can be assigned to when via an object spread  (https://github.com/apple/pkl/pull/428[#428]).
* Relative globbed reads resolve to the same value in different modules (https://github.com/apple/pkl/pull/449[#449]).
* Performance bug: globbed imports and globbed reads expand to an unbounded number of root nodes (https://github.com/apple/pkl/pull/449[#449]).
* Relative globbed imports within a package match no modules (https://github.com/apple/pkl/pull/496[#496]).
* Constraints within typealiases resolve to the wrong values (https://github.com/apple/pkl/pull/144[#144]).
* Members of `pkl.reflect` cannot be rendered (https://github.com/apple/pkl/pull/170[#170], https://github.com/apple/pkl/pull/470[#470]).
* Throws `NullPointerException` if "List Resources Response" or "List Modules Response" messages contain both null `pathElements` and `error` (https://github.com/apple/pkl/pull/480[#480]).
* Classes of `com.oracle.truffle` are not shaded in pkl-config-java-all and pkl-tools (https://github.com/apple/pkl/pull/238[#238]).
* Throws `PklBugException` when running publish check on an invalid URL (https://github.com/apple/pkl/pull/441[#441]).

== Contributors [small]#🙏#

> We would like to thank the contributors to this release (in alphabetical order):

Reviewer 2: Looks like I'm missing here since @bioball landed the final PR for scheme agnostic projects

Author: Ah, apologies! Will make sure to add you.

Author: Added

* https://github.com/fumiya-kume[@fumiya-kume]
* https://github.com/garysassano[@garysassano]

> * https://github.com/hoxell[@hoxell]

Author: ```suggestion
* https://github.com/HT154[@HT154]
* https://github.com/hoxell[@hoxell]
```
